tbmalt/physics/dftb/gamma.py

```diff
@@ -180,6 +196,22 @@ def dri(t, ind):  # Abstraction of lengthy double interleave operation
     gamma[mask] = gamma[mask] / r[mask]  # <─┘
     gamma.diagonal(0, -2, -1)[:] = hubbard_Us[:]  # (eq. 31)
 
+    # off-diagonal values of on-site part for shell resolved calc
+    if basis.shell_resolved:
+        mask_shell = torch.ones_like(gamma, dtype=dtype, device=device).bool()
+        mask_shell.diagonal(0, -2, -1)[:] = False
+        mask_shell = mask_shell * r.eq(0)
+        ua, ub = hubbard_Us[..., mask_shell.nonzero()[..., 0]], \
+            hubbard_Us[..., mask_shell.nonzero()[..., 1]]
+        mask_diff = (ua - ub).abs() < 1E-8
+        gamma_shell = torch.zeros_like(ua, dtype=dtype, device=device)
+        gamma_shell[mask_diff] = 0.5 * (ua[mask_diff] + ub[mask_diff])
+        if torch.any(~mask_diff):
+            ta, tb = 3.2 * ua[~mask_diff], 3.2 * ub[~mask_diff]
```

Comment:
What is the factor of 3.2 doing here? (mostly out of curiosity than anything else)

Reply:
Hi Adam, this factor here is to deal with the on-site terms when the shell pairs have different Hubbard U values (eg., in the mio set, Us and Up are different for C). So the codes here can make sure it behaves the same as DFTB+ (please see line 127 to 137 in shortgammafuncs.F90).